toplevel/topdirs.ml

```diff
@@ -384,11 +384,35 @@ let () =
     )
     "Print the signature of the corresponding value."
 
+let is_nonrec_type id td =
+  let recursive_use = ref false in
+  let nonrecursive_use = ref false in
+  let it_path = function
+    | Path.Pident id' when Ident.name id' = Ident.name id ->
+        if Ident.same id id' then
+          recursive_use := true
+        else
+          nonrecursive_use:= true
+    | _ -> ()
+  in
+  let it =  Btype.{type_iterators with it_path } in
+  let () =
+    it.it_type_declaration it td;
+    Btype.unmark_iterators.it_type_declaration Btype.unmark_iterators td
+  in
+  match !recursive_use, !nonrecursive_use with
+  | _, false | true, _ -> Trec_first
+  | false, true -> Trec_not
```

Comment:
The logic in this function is not obvious because it is aiming not only for correctness, but also for an idiomatic output. For correctness it would suffice to detect recursive uses, and use `nonrec` otherwise. But you want to keep the more common `type t = ...` form in non-recursive cases where `nonrec` is not necessary. You should explain this in a comment, so that people are mentally prepared for the extra work of maintaining `nonreecursive_use` and the delicate dance of boolean pair matching.